### DISCUSSION TITLE
How to add security headers or security library such as Helmet.js

### DISCUSSION THREAD

**Discussion by danielrjames** [nuxt/nuxt#17336] [Questions] [answered]
Hi, I was looking at the nuxt 2 module [nuxt-helmet code](https://github.com/victor-perez/nuxt-helmet/blob/master/lib/module.js) and was wondering how to implement in nitro.

I tried the following similar approach in a local module but had zero results.

```import { defineNuxtModule } from '@nuxt/kit';
import helmet from 'helmet';

export default defineNuxtModule({
  setup(_moduleOptions, nuxt) {
    nuxt.hook('render:setupMiddleware', (app: any) => {
      app.use(helmet());
    });
  },
});
```

Any idea how to add Helmet.js to all ssr pages? Or if that is not an option, a way to manually be able to add the custom security headers?

**Comment by wobsoriano** [comment, marked as answer]
Have you tried creating a server middleware and exporting helmet?

```ts
// ~/server/middlware/helmet.ts
import helmet from 'helmet'
export default helmet()
```

For Nuxt RC.12 and above, you need to wrap it in a  `fromNodeMiddleware` function:

```ts
import { fromNodeMiddleware } from 'h3'
import helmet from 'helmet'

export default fromNodeMiddleware(helmet())
```

**Reply by danielrjames** [reply]
That didn't work. I solved this by manually adding the helmetjs sercurity headers in server middleware though.

```
export default defineEventHandler((event) => {
  appendHeader(event, 'Cross-Origin-Embedder-Policy', 'unsafe-none');
  appendHeader(event, 'Cross-Origin-Opener-Policy', 'same-origin');
  appendHeader(event, 'Cross-Origin-Resource-Policy', 'same-origin');
  appendHeader(event, 'Expect-CT', 'max-age=86400');
  appendHeader(event, 'Referrer-Policy', 'no-referrer');
  appendHeader(
    event,
    'Strict-Transport-Security',
    'max-age=31556926; includeSubDomains; preload'
  );
  appendHeader(event, 'X-Content-Type-Options', 'nosniff');
  appendHeader(event, 'Origin-Agent-Cluster', '?1');
  appendHeader(event, 'X-DNS-Prefetch-Control', 'off');
  appendHeader(event, 'X-Download-Options', 'noopen');
  appendHeader(event, 'X-Frame-Options', 'SAMEORIGIN');
  appendHeader(event, 'X-Permitted-Cross-Domain-Policies', 'none');
  appendHeader(event, 'X-XSS-Protection', '0');
});
```

**Reply by wobsoriano** [reply]
Which part didn't work? I tried just now and it works

```ts
// ~/server/middleware/helmet.ts
import helmet from 'helmet'

export default helmet({
  contentSecurityPolicy: false
})
```

<img width="459" alt="Screen Shot 2022-06-03 at 7 33 07 AM" src="https://user-images.githubusercontent.com/13049130/171875353-251abaf2-b7dd-42d8-a41b-94df87664e94.png">

**Reply by danielrjames** [reply]
Weird. I just tried it again and it worked. Thank you!

**Reply by alpeshvmodi** [reply]
Can someone please suggest how to achieve this in Nuxt 3?

**Reply by viii47** [reply]
@alpeshvmodi you need to install `helmet` and create a new file in `~/server/middleware/`, or put it in one that you already created and add the corresponding code from @wobsoriano

**Reply by alpeshvmodi** [reply]
@viii47, that worked. You really saved me.
Thanks,

**Reply by viii47** [reply]
Happy to hear that it worked 😊
